hubble/cmd/common/k8s/client.go

```diff
@@ -0,0 +1,142 @@
+package k8s
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/spdy"
+
+	"github.com/cilium/cilium/pkg/safeio"
+)
+
+// Slim version of: https://github.com/cilium/cilium/tree/v1.17.0-pre.1/cilium-cli/k8s/client.go
```

Comment:
Can refactor the Cilium CLI and Hubble CLI to share common code for the port-forwarding and service retrieval? Then in a follow up PR we can switch both to "native" port-forwarding instead of exec.

Reply:
That works yes. Looking through the repo, seems `cilium/pkg/hubble` would be a good place to put this common logic (something like `func RelayPortForward(...)`), and have a small k8sclient interface for GetService there, so each client can use its own k8s client (I don't think we could re-use the same client as cilium-cli without tying ourselves to cilium-isms/helm configs/etc).

While at it, I could also update UIPortForward in hubble to re-use that as well instead of hardcoding port 80.

wdyt ?

Reply:
Hmm. ideally we could share the implementation for getting services, and eventually the implementation for `ProxyTCP`. I think it probably isn't a big deal, most of the logic in `cilium-cli/k8s` looks like it isn't too specific to Cilium CLI.  There is quite a bit of logic we aren't going to use, but I don't think that's too big of a deal, the same is true for `client-go` in general. The question is how much work is it going to be to move that package somewhere else, and where to move it. Maybe we can get input from @kaworu or @rolinh on this.

Additionally, it seems like a good portion of Cilium CLI is already using "native" (non-exec) based port forwarding, it's just `cilium port-forward` that isn't using it, funnily enough. Sysdump uses native port-forward for a bunch of stuff already it looks like. So we could actually even switch to native port-forwarding for this initial PR I think without it being a big deal.

Reply:
Also @michi-covalent maybe you know about sharing the k8s related code between cilium CLI and Hubble CLI.

Reply:
I opened a PR to extract the k8s client package out of cilium-cli: https://github.com/cilium/cilium/pull/35482

and opened a PR that uses native port-forwarding in both hubble and cilium-cli: https://github.com/cilium/cilium/pull/35483

I think it turned out pretty well, let's follow-up there :)